### DISCUSSION TITLE
ICYStream with hiccups

### DISCUSSION THREAD

**Discussion by aanban** [pschatzmann/arduino-audio-tools#441] [Q&A] [answered]
Salut Phil, 

congratulations, these are great libs. 

Right now I'm playing around with these two libs (VS-Code + Platformio + ESP32-A1S audio-kit). 
- audiokit
- audio-tools 

The basics are working, i.e. the compiler is happy, the application can be flashed and the hardware works so far, the "sinus" test application provides the 1kHz audio at the headphone output as expected. My project idea is an ICYStream based webradio, finally in the end a player alla ESP32/Vue Radio Player. So to get familiar I compiled the simple ICYStream example and flashed it to the ESP32-A1S based audio kit. Everything seems to run normally, the URLs from your example are loaded and in principle the stream is playing somehow. However only short parts are played and interrupted by a pause. It feels like this is about 500ms sound and 100ms pause. 

I have also read through wiki Performance page and changed the various parameters such as ``DEFAULT_BUFFER_SIZE``, ``I2S_BUFFER_SIZE`` and ``I2S_BUFFER_COUNT``, or activating the task-based Stream but without success. Setting the ``AudioLogger::Info`` or ``Debug`` print more infos to the logfile, but -- as expected -- the sound gets even more interrupts. 

What can be the cause for the hiccups ?
Do I need another setting in platformio.ini, e.g. deactivate the PSRAM or so?
Or anything else to be aware off?

Second issue:
I tested two other URLs ``https://stream.srg-ssr.ch/rsc_de/aacp_96.m3u`` and ``http://liveradio.swr.de/sw890cl/swr3/play.m3u`` to the list below. But it turned out that they do not work, they do not connect properly. What's wrong?

This is my test environment:
```
Windows 10 21H2 (Build 19044.2130)
VS-Code          1.73.0
platformio home  4.3.4
platformio core  6.1.6a1 
espressif 32     5.2.0    
    --> ESP-IDF  4.4.2
    --> Arduino  2.0.5

AUDIOKIT_BOARD=6 ESP32-A1S ai_thinker (AC101) 2957
```

### main.cpp, the simple ICYStream example
```c
#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioKit.h"

const char *urls[] = {
  "https://stream.srg-ssr.ch/rsc_de/aacp_96.m3u",
  "http://liveradio.swr.de/sw890cl/swr3/play.m3u",
  "http://stream.srg-ssr.ch/m/rsj/mp3_128",
  "http://stream.srg-ssr.ch/m/drs3/mp3_128",
  "http://stream.srg-ssr.ch/m/rr/mp3_128",
  "http://sunshineradio.ice.infomaniak.ch/sunshineradio-128.mp3",
  "http://streaming.swisstxt.ch/m/drsvirus/mp3_128"
};
const char *wifi     = "theWlan";
const char *password = "thePass";

ICYStream urlStream(wifi, password);
AudioSourceURL source(urlStream, urls, "audio/mp3");
AudioKitStream kit;
MP3DecoderHelix decoder;
AudioPlayer player(source, kit, decoder);

void next(bool, int, void*) {
   player.next();
}

void previous(bool, int, void*) {
   player.previous();
}

void stopResume(bool, int, void*){
  if (player.isActive()){
    player.stop();
  } else{
    player.play();
  }
}

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}


// Arduino setup
void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Error);

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = false;
  kit.begin(cfg);

  // setup navigation
  kit.addAction(PIN_KEY1, next);
  kit.addAction(PIN_KEY3, previous);


  // setup player
  player.setMetadataCallback(printMetaData);
  player.setVolume(0.7);
  player.begin();
}

// Arduino loop
void loop() {
  player.copy();
  kit.processActions();
}

```
### platformio.ini
```
[env:esp-wrover-kit]
platform = espressif32
board = esp-wrover-kit
framework = arduino

lib_deps = 
  https://github.com/pschatzmann/arduino-audiokit-hal
  https://github.com/pschatzmann/arduino-audio-tools
  https://github.com/pschatzmann/arduino-libhelix

lib_ldf_mode = deep+

; ### PSRAM workaround for Ardunio platform, and debug messages
build_flags =
  -DBOARD_HAS_PSRAM
  -mfix-esp32-psram-cache-issue
  -DUSE_HELIX
  -DAUDIOKIT_BOARD=6

; ### serial monitor setting
monitor_filters = esp32_exception_decoder, log2file
monitor_speed = 115200
monitor_port = COM5

; ### jtag debug interface setting
debug_tool = esp-prog
debug_init_break = tbreak setup
debug_speed = 16000

board_build.partitions = huge_app.csv

```

**Comment by pschatzmann** [comment]
PSRAM is not used by the audiotools library unless you add the related activation class explicitly.
Make sure that the CPU frequency is at max speed. I don't know what PlatformIO proposes as default.
Prefer to use urls with HTTP: HTTPS adds a lot of overhead.
Is your device at a location with good WIFI reception ?

I am not sure why you have this problem. So I suggest the following:
- Try the sketch with Arduino do identify if the issue is related to PlatformIO. 
- All the urls are located in Switzerland, so if you are far, try to use some local URLs: It could be that your network is just too slow - though I was using a mobile phone connection to do my tests...
- You could replace ICYStream with URLStream to save a little bit of processing time - but I don't think that this will  make a big difference
- You can try to activate the USE_URLSTREAM_TASK - though this does not help if your network thruput is too slow and I think we should find the root cause first

**Reply by pschatzmann** [reply]
ps. I just tried in Arduino and I can reproduce the problem when the dip switches were in the debugging position !
https://github.com/pschatzmann/arduino-audiokit/wiki/Problems-with-the-SD-Drive-or-some-conflicting-Buttons

To test I suggest that you lower the complexity and use https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-url_mp3-audiokit/streams-url_mp3-audiokit.ino instead

**Reply by aanban** [reply]
Kewl, thank you for the super fast response. Yes I'm not planing to use the SD card, so I prefer having the JTAG pins free for debugging. OK next tests will be to switch off the SD related things in the config and double check the CPU frequency settings. The WLAN is quite OK. I have another sketch running for the ESP32-A1S Kit and and most URLs run with it. There are very rarely slow streams, and mostly just a reset, or reestablish the connection is the solution. In principle, it runs for hours.

**Reply by pschatzmann** [reply]
Just comment out the   kit.processActions();
This will deactivate the logic to interpret the key presses...

**Reply by aanban** [reply]
OK, I got a kind of interrupt right now as well, sometimes life cames in between..., so back to my tests. 

I double checked the frequency, I added ``dump_eps32_infos()`` that is called at setup time and shows the detail of the running ESP32 device. So from this side everything is fine, the CPU runs fullspeed, the RAM and FLASH are in "green" range.

```
ESP32 Chip ID=12583920
CPU cores = 2
WiFi / BT / BLE
ESP IDF version = v4.4.2 
Silicon revision = 1 
Flash = 4MB external
CPU-Freq = 240 MHz
HEAP  = 83% free of 303820 
PSRAM = 100% free of 4192139 
FLASH = 75% free of 4194304 
```

```c
void dump_eps32_infos(void) {
  uint32_t chipId = 0;
  esp_chip_info_t chip_info;
  esp_chip_info(&chip_info);
  for(int icnt=0; icnt<17; icnt=icnt+8) {
    chipId |= ((ESP.getEfuseMac() >> (40 - icnt)) & 0xff) << icnt;
  }
  
  printf_P(PSTR("\x1B[2J\x1B[1;1HESP32\nChip ID=%d\nCPU cores = %d\nWiFi %s %s\n"), 
                        chipId, 
                        chip_info.cores,
                        (chip_info.features & CHIP_FEATURE_BT) ? "/ BT" : "",
                        (chip_info.features & CHIP_FEATURE_BLE) ? "/ BLE" : "");
  
  printf_P(PSTR("ESP IDF version = %s \nSilicon revision = %d \n") ,esp_get_idf_version(), chip_info.revision);
  printf_P(PSTR("Flash = %dMB %s\n"), spi_flash_get_chip_size() / (1024 * 1024),(chip_info.features & CHIP_FEATURE_EMB_FLASH) ? "embedded" : "external");
  printf_P(PSTR("CPU-Freq = %d MHz\n"), ESP.getCpuFreqMHz());
  printf_P(PSTR("HEAP  = %d%% free of %d \n"), (100 * ESP.getFreeHeap())        / ESP.getHeapSize(),     ESP.getHeapSize());
  printf_P(PSTR("PSRAM = %d%% free of %d \n"), (100 * ESP.getFreePsram())       / ESP.getPsramSize(),    ESP.getPsramSize());
  printf_P(PSTR("FLASH = %d%% free of %d \n"), (100 * ESP.getFreeSketchSpace()) / ESP.getFlashChipSize(),ESP.getFlashChipSize());
}




```
### Next test: turn off ``kit.processAction()`` 
Hooray, as far as I see the kit.processAction() seems to take too much CPU load, once I deactivated it, everything seems to work, at least I got the Jazz Radio on air 👍 

```
ESP32 Chip ID=12583920
CPU cores = 2
WiFi / BT / BLE
ESP IDF version = v4.4.2 
Silicon revision = 1 
Flash = 4MB external
CPU-Freq = 240 MHz
HEAP  = 83% free of 303820 
PSRAM = 100% free of 4192139 
FLASH = 75% free of 4194304 
...........
==> Genre: Entertainment
==> Description: SwissJazz_MP3@128Kbps
==> Name: /SwissJazz_Room1
==> Title: Dave Ruosch Trio - Way Back Blues

```

### Addendum: disabling the SD in the header file did not change anything. 
```c
// This reserves a SPI port
#ifndef AUDIOKIT_SETUP_SD
#define AUDIOKIT_SETUP_SD 0
// for AUDIOKIT_SD_SPI you can select SPI or SPI_VSPI
#define AUDIOKIT_SD_SPI SPI
#endif
```
As fas as I see it, the SD module has been disabled in the main.cpp as well, i.e. the SD module wasn't involved at all.

```c
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = false;
  kit.begin(cfg);

```

**Comment by pschatzmann** [comment, marked as answer]
The issue is not CPU, but a pin conflict:
#define PIN_KEY2 13
Key 2 is assigned to some functionality which is triggered, when you do not deactivate the JTAG and SD

**Reply by aanban** [reply]
OK, your are right, it was my fault. Since I do not use the buttons key2,key4,key5 (I removed the resistors R66, R68 and R69) since I needed free GPIOs for SDA(18) and SCL(23) to get the I2C running for other things. As far as I see I can deaktivate single Keys as well, right?

**Comment by pschatzmann** [comment]
kit.audioActions().setEnabled(int pin, bool enabled) should do the trick...

**Reply by aanban** [reply]
great, thanks again.

**Comment by aanban** [comment]
The unused buttons can be switched off with this function. So hopefully I remember next time :-)

However, I still have a question. 
I saw that also tested the lyrat_v4_3 board from Espressif and added it to the audiokit lib. I took a look at the data sheet and the schematic. In comparsion to the ESP32-A1S kit, I find that the GPIOs are arranged a bit more sensibly, and are available on headers, e.g. I2C, JTAG, and UART. Also the Wrover module is placed in the upper right corner as suggested in the Espressif guidelines. Actually, the WLAN reception should be a bit better. 

What are your experiences when you compare the two audio kits? Which is your favorite?

**Comment by pschatzmann** [comment]
The LyraT does not have the microphone bug and you can select properly between Aux in and Microphones.
If I remember right it is a bit more difficult to program and you need to connect it with 2 cables to do so...

So for anything not needing Aux in I still prefer the AudioKit
